test-app/tests/rendering/action-test.gts

```diff
@@ -0,0 +1,224 @@
+import Helper from '@ember/component/helper';
+import { fn } from '@ember/helper';
+import { on } from '@ember/modifier';
+import { getOwner } from '@ember/owner';
+import { click, render } from '@ember/test-helpers';
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+
+import { action, command } from 'ember-command';
+
+import type CounterService from 'test-app/services/counter';
+
+const inc = action(({ services }) => {
+  console.log('factory called');
+
+  return (amount: number) => {
+    console.log('inc', amount);
+
+    services.counter.counter += amount;
+  };
+});
+
+const incOne = action(({ services }) => {
+  console.log('factory called');
+
+  return () => {
+    console.log('inc');
+
+    services.counter.counter += 1;
+  };
+});
+
+// Invocation Styles:
+//
+// 1. (action)
+//
+// - `<button {{on "click" (incOne)}}>Inc</button>`
+// - `<button {{on "click" (inc 10)}}>Inc</button>`
+// - `<Step @inc={{10}} @push={{(inc)}}>Inc</Step>`
+//
+// 2. (fn (action))
+//
+// - `<button {{on "click" (fn (incOne))}}>Inc</button>`
+// - `<button {{on "click" (fn (inc 10))}}>Inc</button>`
+// - `<button {{on "click" (fn (inc) 10)}}>Inc</button>`
+// - `<Step @inc={{10}} @push={{(fn (inc))}}>Inc</Step>`
+//
+// 3. (command action)
+//
+// - `<button {{on "click" (command incOne)}}>Inc</button>`
+// - `<button {{on "click" (command (inc 10))}}>Inc</button>`
+// - `<Step @inc={{10}} @push={{command inc}}>Inc</Step>`
```

Comment:
These are some _API Design_ ideas. Please give your feedback.

They are declarative so far, but I think, they can be turned imperative, too - somewhat around these lines:

```ts
// ---
const add = action(({ services }) =>
  (a: number, b: number) => services.adder.add(a, b)
);

const doAdd = add(owner);
doAdd(3, 5);

const addOne = (b: number) => doAdd(1, b);
addOne(5);
```

so basically, `action()` returns a factory, than the final function itself - is there a known way around this? 😇

Reply:
how else would you get the owner in there? it's gotta be wired in from either the user, or the manager infra

Reply:
In my defense: The imperative API wasn't there from the start at all. I only focused on the declarative API as its main citizen, but realized, this can be done, too.

I'm less concerned about connecting it with the owner (that's as you said provided or manager infra).

What I'm most interested in (under the terms of what's possible given ember's API):
Does it "feel" good to be used or is it "clumsy"/stiff, rather than inviting to be used? (for both imperative and declarative API)